Update from upstream

perrettecl opened the pull request.

maekitalo committed:
Ssl certificates can be loaded and are used.
The http client can talk with the server https.
This is tested with the demos rpcserver and rpcechoclient.

maekitalo committed:
use std::string instead of const char* to pass certificates to http server.

maekitalo committed:
jsonrpc server and client can use ssl

maekitalo committed:
Binary rpc server and client can use ssl now.
I changed the api a bit. The ssl flag is now passed with a separate method
in json rpc and binary rpc. Also the domain in binary rpc is moved to a
separate method instead of optional parameter to the constructor.

maekitalo committed:
extend rpc benchmark programs with support for ssl

maekitalo committed:
Fix unittest: uri should use json and bin protocol for json and binary rpc.
The protocol field is now verified in the server and hence now the unittest failed.

maekitalo committed:
Merge branch 'ssl'

maekitalo committed:
make ssl support optional with configure switch --with-ssl=true|false

maekitalo committed:
minor fixes
* make constructor of json and binary rpcserver with a event loop explicit
* initialize members when cxxtools::net::Uri is default constructed

maekitalo committed:
make setting ssl just a little nicer.

maekitalo committed:
add missing includes in service registry

maekitalo committed:
prepare for version 3.0rc1

maekitalo committed:
remove use of cxxtools::NonCopyable and make it deprecated and other minor code cleanup

maekitalo committed:
fix potential rounding problem in unit test of serialization info

maekitalo committed:
return reference to argc in application class for better support for cxxtools::Arg

maekitalo committed:
add missing helper methods queueEvent and commitEvent to application class
They were documented but not implemented.

maekitalo committed:
queuePriorityEvent in application class was delegated to the wrong method (commitPriorityEvent) - fixed.
Also the documentation for queuePriorityEvent was wrong.

maekitalo committed:
Log initialization change.
Logging is not (re-)initialized with a default constructed log configuration.

maekitalo committed:
small documentation fix

maekitalo committed:
add a prefix for silent rules for automake

maekitalo committed:
remove obsolete "register" declaration

maekitalo committed:
add signal to event loop when the loop is started

maekitalo committed:
Improve serialization of empty objects with just a type name.

maekitalo committed:
make base64 encoder configurable

maekitalo committed:
complete FileDevice

maekitalo committed:
simplification of io device by providing default implementations of most virtual methods.

maekitalo committed:
implement cxxtools::IODevice for standard input, output and error

maekitalo committed:
fix stddevice and add demo

maekitalo committed:
fix compile problem - forgot to fix filedevice.h

maekitalo committed:
implement serialization and deserialization operators for long double

maekitalo committed:
we do not need to call fsync on the wake pipe since pipes do not need fsync

maekitalo committed:
When accept4 is used, the close on exec flag is set by that call. Subsequent setting of the flag is obsolete then.

maekitalo committed:
call fsync just on file devices

maekitalo committed:
fix some error strings

maekitalo committed:
Merge branch 'iodevice'

maekitalo committed:
Extend cxxtools::SerializationInfo by float and double types.
This reduces rounding problems due to converting float and double values to long double.

maekitalo committed:
Extend insrv demo with network server functionality.
The demo shows how to implement asyncronous I/O using cxxtools.

maekitalo committed:
add asyncronous functionality to cxxtools::IOStream

maekitalo committed:
add some documentation about new insrv demo

maekitalo committed:
add asyncronous functionality to cxxtools::IStream and cxxtools::OStream

maekitalo committed:
implement input stream to stdin and output stream to stdout and stderr

maekitalo committed:
fix dist package: add cxxtools/stdstream.h

maekitalo committed:
fix initialization of cxxtools::StdoutDevice and cxxtools::StderrDevice

maekitalo committed:
Limit optionally the number of events processed per cycle in event loop

tommi-maekitalo committed:
make hexdump leaner and easier

tommi-maekitalo committed:
use new api for hexdump in hd demo; add missing include <cxxtools/hexdump.h> to dist package

maekitalo committed:
add binary of new demo to .gitignore

maekitalo committed:
Merge branch 'master' into 3.0

Commit:
fix handling poll when POLLIN and POLLHUP is both set

maekitalo committed:
update version number

maekitalo committed:
Fix retrieving floating point value in SerializationInfo.
It was wrong in some circumstances after extending the union with float and double

maekitalo committed:
Fix handling NaN values in json and binary serialization.
We need special handling which was partly lost after extending the union of SerializationInfo

maekitalo committed:
fix building without C++11 support

maekitalo committed:
Fix reopening sockets while handling I/O events.
Sometimes the close event is emitted although the socket was closed and reopened.

maekitalo committed:
fix crash in debug output, which was added by last commit

maekitalo committed:
fix typo in debug output

maekitalo committed:
fix serialization of arrays with type name in binary serialization

tommi-maekitalo committed:
fix in binary deserializer

dictionary was not cleared when deseriaizer was reused

maekitalo committed:
fix in binary rpc

The last fix broke binary rpc.
Make resetting dictionary optional since it must not be done in binary rpc
when multiple arguments are deserialized.

tommi-maekitalo committed:
define serialization operator for cxxtools::QueryParams

Commit:
fix checking error in tcp client socket

When on blocking connect when the first address from getaddrinfo fails and
the second do not succeed successfully immediately (EOINPROGRESS) the tcp
client should wait but it throwed the postponed error message from the first
try.

Commit:
add info log entry in log file when file is rotated

tommi-maekitalo committed:
fix binary rpc server

The dictionary was not cleared between 2 rpc calls which lead in some circumstances wrong objects.

maekitalo committed:
fix automatic reconnect on blocking calls in binary and json rpc client

Commit:
Merge remote-tracking branch 'remotes/origin/3.0'

maekitalo committed:
fix overrun handling when adding timespan to datetime and a day boundary is hit

maekitalo committed:
fix connecting tcp socket in async mode

The connection failed when the first address got from getaddrinfo failed
event when there were more addresses available. When more addresses
are available, it should just try the next one.

adig33 committed:
imported from db879

maekitalo committed:
Implement support for ssl client certificates

maekitalo committed:
replace try-catch cleanup with temporary object with destructor in event loop

maekitalo committed:
Optimization of binary deserialization
Instead of passing characters one by one to parser pass a streambuf and let the
parser consume as many characters it needs or until the streambuf is empty.
Measures using serializer-bench tool indicates a 20-25% performance improvement.

maekitalo committed:
Fixes for asyncronous ssl in tcp socket

maekitalo committed:
initialize openssl library just once

Commit:
Merge branch 'master' of github.deutsche-boerse.de:dg879/cxxtools

maekitalo committed:
Merge remote-tracking branch 'dev/master'

maekitalo committed:
fix parsing csv data with delimiter inside

Commit:
Merge branch 'master' of github.deutsche-boerse.de:dg879/cxxtools

maekitalo committed:
Another small optimization in binary parser

maekitalo committed:
Merge branch 'master' of github.deutsche-boerse.de:dev/cxxtools

Commit:
Merge branch 'master' of github.deutsche-boerse.de:dg879/cxxtools

maekitalo committed:
binary deserializer: fail when trying to deserialize from memory when more data is available after deserialization

Commit:
Merge branch 'master'

maekitalo committed:
fix binary serialization of zero characters

maekitalo committed:
Optimize event loop

The event loop maintains now a list of actively processed events. The list is accessed
only by the current thread and hence do not need to be locked. So when
multiple events are in the loop the lock need to be aquired just once

maekitalo committed:
add saml2 demo to gitignore

maekitalo committed:
add a log debug statement when a new binary rpc request starts

maekitalo committed:
fix binary formatter
binary formatter did not handle values without type name correctly

maekitalo committed:
Timer improvements
Allow passing date and time as gmt and implement callback signal, which
receives due time in gmt.

maekitalo committed:
fix example of array in documentation of binary serialization

maekitalo committed:
fix possible overflow in cxxtools::Time::setTotalMSecs

maekitalo committed:
Fix in Timer: once timer (after, at) can now be restarted in callback

maekitalo committed:
Timer: rename gmt to utc

maekitalo committed:
support special characters in query params when deserializing into SerializationInfo

maekitalo committed:
last fix in query params was not correct - now fixed

maekitalo committed:
always set TCP_NODELAY
The socket option disables the nagle algorithm which is normally useless.
In cxxtools std::streambuffer is used to build larger network frames to
send and if the application flushes the buffer, the tcp packet should be
sent immediately instead of waiting for more data to be send.
This speeds up larger rpc calls significantly.

maekitalo committed:
bugfix: the TcpSocket needs to check if the device is closed after checking the poll events
and emit the `closed` signal then.

maekitalo committed:
Unify setting selector with setSelector.
Previously sometimes a selector must be passed via pointer and sometimes via reference.
Now both is allowed everywhere.

maekitalo committed:
Demo insrv must react on closed signal from tcp socket.
Also some comments are added to make the demo easier to understand.

maekitalo committed:
remove now obsolete query params unit test

maekitalo committed:
add flag -T <timespan> to rpcbenchclient for maximum runtime of rpc test

maekitalo committed:
fix memory leak when reading peer ssl certificates

maekitalo committed:
Ssl certificate subject and issuer is now returned as unicode string.
Also the deprecated openssl function X509_NAME_online is replaced by newer API.

maekitalo committed:
add methods to return validity dates for ssl certificates

maekitalo committed:
add ssl signals to TcpStream

maekitalo committed:
add delegate for verifying ssl certificate

maekitalo committed:
Fix http, json and http servers
When accept call failed, the servers stopped responding.

maekitalo committed:
http, json and binary rpc servers do not block any more while ssl handshake.

maekitalo committed:
new class cxxtools::SslCertificate

maekitalo committed:
Change binary rpc server api.

Instead of loading a ssl certificate in a separate call, it is given as
a parameter to the listen call. The certificate is now specific to a listener
and not to the server any more. Since multiple listeners can be instantiated
by calling listen multiple times, it is possible to listen on ssl and non
ssl in the same server or with different certificates on different interfaces.

The backlog parameter is removed and set to 64.

maekitalo committed:
Implement same api change in json rpc server as in binary rpc server.

Now also the json rpc server can listen on ssl and non ssl connections at the
same time.

maekitalo committed:
Implement same api change in http server as in binary and json rpc server.

Now also the http server can listen on ssl and non ssl connections at the
same time.

maekitalo committed:
remove some now unused variables in rpc servers

maekitalo committed:
change the way, how to activate ssl in binary rpc client
Ssl options are now passed with connection parameters. Either just a flag
to activate ssl without client certificate or a certificate file name
to set a client certficate.

maekitalo committed:
improve ssl client certificate support.
- The delegate `acceptSslCertificate` is now available in binary rpc client
- Clearify documentation about ssl certificate verification
- Load certificate from file
- Add getter for serial number of certificate

maekitalo committed:
add acceptSslCertificate delegate to httpclient and json rpc client

maekitalo committed:
change activating ssl in json client as previously done in binary rpc client

maekitalo committed:
add support for client side certificates to http client

maekitalo committed:
fix ssl client certificate handling in binary rpc client

maekitalo committed:
fix ssl client certificate handling in json rpc client

maekitalo committed:
add `acceptSslCertificate` delegate to binary and json rpc server

maekitalo committed:
implement client certificates in http client and server and json http rpc client

maekitalo committed:
add some demo/test programs for client certificates

maekitalo committed:
implement client side certificates in xmlrpc client

maekitalo committed:
add missing Makefile.am for new ssl demos

maekitalo committed:
add some convenience methods for asyncronous ssl to TcpStream

maekitalo committed:
fix building with openssl < 1.0.2
We use ASN1_TIME_diff to convert validity dates of certificates to DateTime.
This is not supported before 1.0.2 so we have to parse the date time string.

maekitalo committed:
setup locking for openssl to be thread safe

maekitalo committed:
rename ssl demos

maekitalo committed:
add support for server certificate verification to rpc clients

maekitalo committed:
fix use ASN1_INTEGER_to_BN (openssl function)
It did not work with openssl 1.1.0

maekitalo committed:
fix copyconstructor of cxxtools::SslCertificate: the passed argument must be const

maekitalo committed:
Fix executing sub processes.
cxxtools::posix::Exec::exec() must not allocate memory. This is not allowed
in multithreading programs since it can't be ensured, that another thread
do not lock the heap after a fork.

maekitalo committed:
Fix executing sub processes.
cxxtools::posix::Exec::exec() must not allocate memory. This is not allowed
in multithreading programs since it can't be ensured, that another thread
do not lock the heap after a fork.

maekitalo committed:
minor fix in error message

maekitalo committed:
ssl handle must be released with SSL_free instead of SSL_clear

maekitalo committed:
fix problems when ssl connection is initiated

maekitalo committed:
catch abi::__forced_undwind on thread end when exists.
This prevents a core when a thread is canced using pthread_cancel in gnu libstdc++

maekitalo committed:
Optimize json serializer
The json serializer wrote the formatted data to a cxxtools::TextStream, which
is a filter and does character conversion (e.g. UTF-8). But since json data
is just ASCII the conversion is unnecessary. Removing the text stream doubles
the serialization speed.

maekitalo committed:
Remove now obsolete character type template parameter from json output object

maekitalo committed:
Binary serializer: fix type of size parameter when deserializing from memory

maekitalo committed:
fix formatting numbers in json formatter
It was broken after last change since the locale of the output stream was
used. This may lead to thousands separators or decimal comma, which is not
allowed in json format.

maekitalo committed:
do not log warning that accept is terminated when rpc servers shut down

maekitalo committed:
small change in serialization: setting null sets the category to Value

maekitalo committed:
Merge branch 'master' of github.deutsche-boerse.de:dev/cxxtools

maekitalo committed:
fix formatting dates with negative year including the default constructed date.

maekitalo committed:
bugfix: make blocking sslConnect, sslAccept and sslShutdown really blocking

maekitalo committed:
directory class can now iterate over directories referenced by links

maekitalo committed:
Improve cxxtools::Arg
* Dereference operator returns now the value
* Arg with char accepts white space

maekitalo committed:
added client certificate checking to jsonrpc over http client

maekitalo committed:
bugfix: do not access TcpSocketImpl when object is deleted in a callback

maekitalo committed:
fix link problem with ssl demos

maekitalo committed:
small fix in unit test: statistics are accumulated when multiple units are executed

maekitalo committed:
small style changes

maekitalo committed:
Merge branch 'master' of github.deutsche-boerse.de:dg879/cxxtools

maekitalo committed:
fix blocking when adding data to a streambuffer, which is already writing

maekitalo committed:
fix blocking when adding data to a streambuffer, which is already writing

maekitalo committed:
Merge branch 'master' of github.deutsche-boerse.de:dev/cxxtools

jwilk committed:
fix typo

maekitalo committed:
improve error reporting in binary deserializer

maekitalo committed:
Fix error checking in iodevice and tcpsocket.
errno must be saved befor logging.

maekitalo committed:
fix reading in streambuffer
buffer was not reinitialized correctly on beginRead when data was already in buffer

maekitalo committed:
fix reading from streambuffer
Under high load the streambuffer may reinitialize the write buffer wrongly.
This happens when data was pending on overflow. The write buffer was initialized
with the old size although some data was already written.

maekitalo committed:
Merge pull request #20 from jwilk-forks/spelling

Fix typo

maekitalo committed:
improve date, time and date time parsing and formatting

* format code %O reads and writes english month names
* format code %N writes day of week as english abbrivation
* parse token '#' skips a word
* when parsing fails mark the position where it failse in the error message with (*)

maekitalo committed:
fix annother errno which may be lost when logging is enabled in tcpsocket

maekitalo committed:
Better handling of multiline values in mime headers

maekitalo committed:
Better handling of system errors while handling ssl connections

Instead of reporting a logic_error on endSsl... when beginSsl... fails report a
IOError with a suitable message.

maekitalo committed:
new demo "siconvert" which converts serialization formats to others

maekitalo committed:
add some useful log debug messages to SerializationInfo

maekitalo committed:
Disable logging after fork.
This prevents a deadlock when the child process tries to log while another thread
of the parent helds the logging lock.

maekitalo committed:
Fix last commit: disable logging after fork only in child process

maekitalo committed:
forgot to add source for new demo siconvert - here it is

maekitalo committed:
fix reading binary data from pipestream: '\xff' was misinterpreted as eof

maekitalo committed:
add method tryLock to class MutexLock

maekitalo committed:
support serialization and deserialization of std::vector<bool>

maekitalo committed:
reduce rounding errors when serializing cxxtools::Timespan classes

maekitalo committed:
make cxxtools compatible with C++03 again

maekitalo committed:
new class cxxtools::PtrStream, which implements a iostream interface to a buffer.

maekitalo committed:
implement serialization of std::tuple

maekitalo committed:
implement serialization of std::array

maekitalo committed:
implement serialzation of c-array

maekitalo committed:
Make explicit cast in cxxtools::Char when moving or copying data
This removes a warning message from the compiler.

maekitalo committed:
Fix serializing timespans to json

Timespans (and other types) must be serialized as numbers instead of strings.
The timespans were serialized as strings for fixing rounding problems. That
broke the serialization in json.

maekitalo committed:
Timespans are now serialized as bcd in binary serializer.
This saves space and time.

maekitalo committed:
SerializationInfo must be null after calling clear.

This was not the case when it held a string.

maekitalo committed:
bugfix in cxxtools::convert
converting an empty string to signed value dereferences the end iterator.

maekitalo committed:
Small optimization of binary serialization format.

Binary serialization format had already a bcd encoding for numbers. Now I
generalizied it and use it for each string, which fits into the format. This
saves space especially for dates, times and datetimes.

barraudl committed:
bugfix : csv deserializer don't understand "" as " and crash

Signed-off-by: barraudl <lilianbarraud@eaton.com>

maekitalo committed:
add unit test for double quote in csv deserializer

maekitalo committed:
Merge branch 'barraudl-master'

maekitalo committed:
fix in http server: event loop was not set in 2 constructors even when it was passed as parameter

maekitalo committed:
Improve auto detection correct category in serializationinfo.

Category was set to "Object" when members were added and it was not "Array".
Now it is set to "Array" when the category is not set and name parameter is empty.
"Object" category need non empty member names.

maekitalo committed:
Make class cxxtools::EnvSubst private and simplify it a little.

The class is used just in cxxtools::envSubst function and it is sufficient
to make that function public.

maekitalo committed:
implement helpers for serializing and deserializing typed enums

maekitalo committed:
add flag -f to siconvert demo for "follow" mode

this reads messages from input stream until eof instead of stopping after the first message

maekitalo committed:
fix dist package

A missing backslash caused, that some files were not in the dist package

maekitalo committed:
fix some comments

maekitalo committed:
Make siconvert a tool instead of just a demo

It is now installed into the bin directory.
The tool allows to convert data between the various formats supported by the
serialization of cxxtools.

maekitalo committed:
improve siconvert tool

Replace flag -f with --skip and --num options.
Now without -f is like "--num 1" and with -f is the default

maekitalo committed:
reduce #includes

maekitalo committed:
change siconvert tool: -N outputs number of records; -n option removed.

-n printed the number of nodes in the first level.

maekitalo committed:
Authors file is UTF-8, so I can fix my name

maekitalo committed:
Add function cxxtools::serialization_cast

It can be used for convenient conversion of objects via serialization.

maekitalo committed:
remove not existing method from class definition

maekitalo committed:
change version number to 3.0rc3

maekitalo committed:
fix integer overflow on 32 bit systems

maekitalo committed:
fix missing include
sleep(3) needs stdlib.h; we replace that sleep with cxxtools::Thread::sleep in the demo

maekitalo committed:
make casting between types explicit where needed for clearification and hence reduce warnings in cxxtools::Char

maekitalo committed:
Add classes for working with time zones (cxxtools::Tz)

maekitalo committed:
fix formatting

maekitalo committed:
fix reading own time zone

maekitalo committed:
use new UtcDateTime and LocalDateTime classes in DateTime class.

DateTime::localtime returns now a LocalDateTime and DateTime::gmtime a UtcDateTime.
Both are derived from DateTime but are not implicit convertable to each other.

maekitalo committed:
fix building without ssl

Ssl support and dependency to openssl can be disabled with the configure flag --without-ssl.
This was broken and lead to a compile error. This is fixed now.

maekitalo committed:
input stream iterator for cxxtools::QueryParams

maekitalo committed:
siconvert tool can now decode query strings

maekitalo committed:
siconvert tool can read "settings" format

Settings format is really deprecated. Should not be used for new code.

maekitalo committed:
demo for async tcp server

maekitalo committed:
minor fix in usage output of siconvert

maekitalo committed:
fix: cxxtools::split must result in a empty result on empty input

maekitalo committed:
fix compile error in new async demo

maekitalo committed:
fix operator== for cxxtools::net::AddrInfo

maekitalo committed:
new functions trimi, rtrimi and ltrimi for trimming strings in place

maekitalo committed:
fix reading unicode from csv

maekitalo committed:
replacement string in regular expression class now allows curly braces around variable number

e.g. foo$1bar is the same as foo${1}bar

maekitalo committed:
minor fix in xml writer: do not write an eol if no xml prefix is written

maekitalo committed:
fix crash in time zone class - timezones without transitions were not handled correctly.
Also a wrong exception type was thrown when time zone file did not have the expected magic.

maekitalo committed:
fix more bugs in timezones without transitions

maekitalo committed:
set version number to 3.0

maekitalo committed:
implement serializer and deserializer for ini format

maekitalo committed:
allow values before sections in ini deserializer

maekitalo committed:
add compatibility methods to cxxtools::IniFile to older version

maekitalo committed:
Improve ini serialization

add serialization of single values to ini file serialization
add unit test for ini file serialization

maekitalo committed:
deactivate example code in demo tool getini

maekitalo committed:
make setSelector in RemoteClient a virtual method.

Now we can set the selector in remote clients without knowing the exact type.

maekitalo committed:
Remove atomicity stuff

Atomicity is part of C++11 and hence we make C++11 now mandatory.

maekitalo committed:
fix compile error in template specialization of ScopedIncrement

maekitalo committed:
ssl fix

When a socket is closed the ssl context needs to be invalidated.

maekitalo committed:
fix handling non ascii strings in binary serialization

maekitalo committed:
support UTF-16 surrogate pairs in json for unicode chars >= 0x10000 according to ecma standard

maekitalo committed:
parsing csv creates now a array of objects instead of array of arrays

maekitalo committed:
implement codec for win-1252

maekitalo committed:
implement command line tool cxxtz for querying time zone database

maekitalo committed:
new static function cxxtools::getTimeZones() and a new command line tool cxxtz to show time zone informations

maekitalo committed:
Improve error message when deserialization fails

maekitalo committed:
beginXXX-methods must not throw exceptions but they must be postponed to the corresponding endXXX

This was not done in ssl operations of tcpsocket. Now that we have C++11
mandatory, we can use the nice std::exception_ptr to implement that easily.

maekitalo committed:
fix comment in cxxtools::EnumClass

maekitalo committed:
fix timed wait for condition

Time was not calculated correctly and hence the wait may return too early

aquette commented:
@perrettecl looking at [this](https://github.com/maekitalo/cxxtools/blob/master/configure.ac#L1), you should import the latest upstream/master as 42ity:3.0-FTY-master (not 2.2) which we then can use for our master builds (to fix), while 2.1.0 will stay with 2.2 ;)

maekitalo committed:
fix string assignment

Assignment was wrong, when the specified end position of the string to assign
was beyond the string to assign. E.g. assigning "abcd" with pos=2 and length=4
should assign just 2 bytes but it assigned 4 by reading beyond the source string.

aquette commented:
@perrettecl my above comment (from Feb 12th) still stands and is still valid ;)

maekitalo committed:
remove superflous explicit assignment operator in cxxtools::Char

maekitalo committed:
fix utf8 codec

In some circumstances like reading using a cxxtools::TextIStream the codec
breaked processing instead of coninuing consuming characters.
